### PR TITLE
Feat: checkpoint optimize function to restart after crash

### LINKED ISSUE #137: Make `optimize` continue from last checkpoint after crash
When running `optimize`, my process somehow crashed after 4h (was estimated to take 10h).
Now I have to restart it from scratch.
Could you add a checkpointing feature such that it automatically continues with the last chunk?

### PR DESCRIPTION
**Pull request by deependujha**
<details>
  <summary><b>Before submitting</b></summary>

- [ ] Was this discussed/agreed via a Github issue? (no need for typos and docs improvements)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/lit-data/blob/main/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure to update the docs?
- [ ] Did you write any new necessary tests?

</details>

## What does this PR do?

Fixes #137 .

Make `optimize function` fault-tolerant and restart from last checkpoint after an unexpected crash. 

The checkpoint feature is supported for `mode = "None | append"`. `Overwrite` mode will overwrite everything, even checkpoints. 

>
> If you were executing **Optimize function in overwrite mode with `use_checkpoint=True`**, and if it crashes.
> **Execute with `use_checkpoint=True & mode=None` in the next run.**
>

- Sample code for this feature
```python

from litdata import optimize, StreamingDataset

def fn(i: int):
    if i in [22, 36, 59, 78]:
        raise ValueError("An error occurred")

    return i, i**2

def another_fn(i: int):
    return i, i**2

output_dir = "s3://my-dummy-bucket-litdata/is-litdata-checkpoint-working/"
output_dir = "data"

try:
    optimize(
        fn=fn,
        inputs=list(range(100)),
        output_dir=output_dir,
        chunk_size=5,
        num_workers=4,
        use_checkpoint=True,
    )
except Exception as _:
    print("fn failed. Now, we will go for another_fn.")
    optimize(
        fn=another_fn,
        inputs=list(range(100)),
        output_dir=output_dir,
        chunk_size=5,
        num_workers=4,
        use_checkpoint=True,
    )

ds = StreamingDataset(output_dir)
print(f"{len(ds)=}")
print(f"{ds[:]=}")
print("All done ✅")
```

---

🚧 Warning: It doesn't support generator functions as of now. It'll be added in the next PR.

## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in GitHub issues there's a high chance it will not be merged.

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: src/litdata/processing/utilities.py [hunk 1]
```diff
@@ -249,3 +249,37 @@ def extract_rank_and_index_from_filename(chunk_filename: str) -> Tuple[int, int]
...
+    """
+
+    if not filepath.__contains__(".checkpoints"):
+        return filepath
+
+    # uuid is of 32 characters, '.json' is 5 characters and '-' is 1 character
+    return filepath[:-38] + ".json"
+
+
+def downloadDirectoryFromS3(bucketName: str, remoteDirectoryName: str, localDirectoryName: str) -> str:
```
Comment (tchaton) [inline src/litdata/processing/utilities.py:270]:
```suggestion
def download_directory_from_S3(bucketName: str, remoteDirectoryName: str, localDirectoryName: str) -> str:
```

File: src/litdata/processing/utilities.py [hunk 2]
```diff
@@ -249,3 +249,37 @@ def extract_rank_and_index_from_filename(chunk_filename: str) -> Tuple[int, int]
...
+    s3_resource = boto3.resource("s3")
+    bucket = s3_resource.Bucket(bucketName)
```
Comment (tchaton) [inline src/litdata/processing/utilities.py:272]: It is good to keep python style to use `bucket_name` vs `bucketName`. This applies in several places.

**Review by tchaton — APPROVED** [review]
Looks great !

File: src/litdata/processing/utilities.py (hunk 2 above)
Comment (deependujha) [reply src/litdata/processing/utilities.py:272]: sure.

**Review by justusschock — APPROVED** [review]

**Pull request merged**